manifests/rule.pp

```diff
@@ -42,23 +42,40 @@
 
   case $::osfamily {
     'RedHat': {
-      file { "rule-${interface}":
-        ensure  => present,
-        owner   => root,
-        group   => root,
-        mode    => '0644',
-        path    => "/etc/sysconfig/network-scripts/rule-${interface}",
-        content => template('network/rule-RedHat.erb'),
-        notify  => $network::manage_config_file_notify,
-      }
-      file { "rule6-${interface}":
-        ensure  => present,
-        owner   => root,
-        group   => root,
-        mode    => '0644',
-        path    => "/etc/sysconfig/network-scripts/rule6-${interface}",
-        content => template('network/rule6-RedHat.erb'),
-        notify  => $network::manage_config_file_notify,
+      if versioncmp($::operatingsystemmajrelease, '8') >= 0 {
+        if ! defined(Concat["/etc/sysconfig/network-scripts/ifcfg-${name}"]) {
+          concat { "/etc/sysconfig/network-scripts/ifcfg-${name}":
```

Comment:
I'd prefer not to change so radically this define, from using a file to using concat. From what I see you do this in order to be able to add the routing rule fragment from the rule define. Why not just adding a parameter to the interface define, and to the relevant template used by the file resource?

Reply:
I thought this would be add better compatiblity with the other approaches creating ``/etc/sysconfig/network-scripts/rule-<dev>`` in network::rule.

If you like I can rework it to just use the interface and the corresponding template instead.
It's up to your decision.
As I already said, I thought this solution would fit better into the current resource layout.
With this solution people don't need to rework their code for rhel8. They just can use the resources as described in the readme etc.

Reply:
@alvagante what do you prefer?

Reply:
@pseiler for this version I'd prefer the approach that reduces as much as possible the impact on existing setups and preserves backwards compatibility

Reply:
so, basicallyt, I'd keep the file resource approach here

Reply:
@alvagante I'm not sure anymore what to do. You're last two comments counter each other in my understanding.

Backward compatiblity would be more garantueed with my approach. If we change this back to a file resource instead of concat::fragments we must add more parameters to the defined resource "network::interface" and break compatiblity with the approach of an additional resource for routing roules ( "network::rules") because rules behave completely different for RHEL >= 8 when using networkmanager than the rest.

f.e. 
with concat fragments:
```puppet
network::interface { 'eth0':
  ipaddress => '192.168.0.10',
  netmask   => '255.255.255.0',
}
network::routing_table { 'vlan22':
  table_id => '200',
}

network::rule { 'eth0':
  iprule => ['from 192.168.22.0/24 lookup vlan22', ],
}
```

with file resource:
```puppet
$my_iprule = ['from 192.168.22.0/24 lookup vlan22']

network::interface { 'eth0':
  ipaddress => '192.168.0.10',
  netmask   => '255.255.255.0',
  iprule => $::operatingsystemmajrelease ? {
    '8'     => $my_iprule,
    default => undef,
  }
}
network::routing_table { 'vlan22':
  table_id => '200',
}
if versioncmp($::operatingsystemmajrelease, '8') = -1 {
  network::rule { 'eth0':
    iprule => $my_iprule,
  }
}
```

So to be fully backward compatible concat::fragments would be the smoother solution imo.
If i missed something please tell me.

Reply:
@pseiler sorry for lack of clarity, I intended to keep oin using the file resource, adding a parameter, optional and undef by default, called iprule to network::interface, that, if present, add an entry to the relevant template. 
Somethinge like this https://github.com/example42/puppet-network/commit/18fb50315bb8db943153028c516425d13fdd018e which doesn't change existing files on existing systems (and doesn't trigger any service restart) if the module is updated, but add the option to specify the rule, when needed.
It's not necessary to add the check on the rhel version, as that's something left to the user (in his own profiles) when declaring the network::interface define

Reply:
Ah now I get it. Because of the resource change (file -> concat) the network will be restarted.
So the code should kinda work like the snippet above with the file approachment. 
I will change this the next days and mention you when I'm done.
Thank you for clearifying this. Looking forward to get this merged.

Reply:
@alvagante please have look now. I changed everything according to your advice.